File: src/Python/101-200/143.ReorderList.py

class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def reverseList(self, head):
        prev = None
        while head is not None:
            next = head.next
            head.next = prev
            prev = head
            head = next
        return prev


    def reorderList(self, head):
        """
        :type head: ListNode
        :rtype: None Do not return anything, modify head in-place instead.
        """
        if head is None:
            return 
        slow = head
        fast = head.next
        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next
        right = self.reverseList(slow.next)
        slow.next = None
        left = head
        while right is not None:
            next = right.next
            right.next = left.next
            left.next = right
            right = next
            left = left.next.next
        return left

if __name__ == "__main__":
    solution = Solution()
    nodes = [ListNode(1), ListNode(2), ListNode(3), ListNode(4), ListNode(5)]
    nodes[0].next = nodes[1]
    nodes[1].next = nodes[2]
    nodes[2].next = nodes[3]
    nodes[3].next = nodes[4]
    print(solution.reorderList(nodes[0]))

File: src/CSharp/101-200/143.ReorderList.cs

using System;
using System.Text;
using System.Collections.Generic;

namespace PeefyLeetCode.ReorderList
{
public class ListNode {
    public int val;
    public ListNode next;
    public ListNode(int x) {
        val = x;
        next = null;
   }
}

public class Solution {
    public void ReorderList(ListNode head) {
                if (head == null || head.next == null)
            return;
        ListNode p1 = head;
        ListNode p2 = head;
        // 找到链表的一半
        while (p2.next != null && p2.next.next != null) {
            p1 = p1.next;
            p2 = p2.next.next;
        }
        // 将链表分为两段
        p2 = p1.next;
        p1.next = null;
        p1 = head;
        // 将后半段进行链表的翻转
        ListNode head2 = p2;
        ListNode next2;
        while (p2.next != null) {
            next2 = p2.next;
            p2.next = next2.next;
            next2.next = head2;
            head2 = next2;
        }
        p2 = head2;
        // 两条链表进行合并
        ListNode next1;
        while (p2 != null) {
            next1 = p1.next;
            next2 = p2.next;
            p1.next = p2;
            p2.next = next1;
            p1 = next1;
            p2 = next2;
        }
    }
}

}


